### PR TITLE
refactor(core): update peerDependencies to allow rxjs7

### LINKED ISSUE #41897: Support for rxjs 7
# Description

Angular 12 workspace comes bundled with rxjs `6.6.0`
```shell
❯ npx @angular/cli@12.0.0-rc.1 new web-apps --create-application false --skip-install
```
angular_12_rxjs_6 6

### Describe the solution you'd like

Angular 12 workspace should come bundled with [rxjs 7](https://www.npmjs.com/package/rxjs)

RxJS 7 Slides: https://docs.google.com/presentation/d/1-LU7YE3NWw8jHeAgdmLu4CBfG7osCx6MsSIeFs16k60

I see pull request https://github.com/angular/angular-cli/pull/20669 was closed recently.

### PR DESCRIPTION
**Pull request by crisbeto**
We can't update the framework to rxjs7 until version 13, because it contains breaking changes, but we can allow users to opt into it since all of our code should be compatible.

These changes expand the allowed version range of rxjs and add an integration test to verify that we don't get compilation errors. Note that we also have a test that runs the AIO examples against rxjs 7 already (#42660).

Fixes #41897.

### CODE REVIEW

File: packages/common/test/pipes/async_pipe_spec.ts [hunk 1]
```diff
@@ -22,7 +22,7 @@ import {Subscribable, Unsubscribable} from 'rxjs';
       // the implementation does not rely on other methods:
       const wrapSubscribable = <T>(input: Subscribable<T>): Subscribable<T> => ({
         subscribe(...args: any): Unsubscribable {
-          const subscription = input.subscribe(...args);
+          const subscription = input.subscribe.apply(input, args);
```
Comment (gkalpak) [inline packages/common/test/pipes/async_pipe_spec.ts:25]: OOC, why is this change necessary?

File: integration/typings_test_rxjs7/package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    "@angular/forms": "file:../../dist/packages-dist/forms",
+    "@angular/platform-browser": "file:../../dist/packages-dist/platform-browser",
+    "@angular/platform-browser-dynamic": "file:../../dist/packages-dist/platform-browser-dynamic",
+    "@angular/platform-server": "file:../../dist/packages-dist/platform-server",
+    "@angular/router": "file:../../dist/packages-dist/router",
+    "@angular/service-worker": "file:../../dist/packages-dist/service-worker",
+    "@angular/upgrade": "file:../../dist/packages-dist/upgrade",
+    "@types/jasmine": "file:../../node_modules/@types/jasmine",
+    "rxjs": "^7.3.0",
+    "typescript": "~4.3.4",
```
Comment (gkalpak) [inline integration/typings_test_rxjs7/package.json:22]: Why not `file:../../node_modules/typescript`?

File: integration/typings_test_rxjs7/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+      "dom",
+      "es2015.collection",
+      "es2015.iterable",
+      "es2015.promise"
+    ],
+    "types": [],
+  },
+  "files": [
+    "include-all.ts",
+    "node_modules/@types/jasmine/index.d.ts"
```
Comment (gkalpak) [inline integration/typings_test_rxjs7/tsconfig.json:24]: OOC, shouldn't this be in the `types` list?

**Review by gkalpak — APPROVED** [review]
Not sure if `feat` is the right commit type. Is that what we've uses for other similar changes in the past?
Otherwise lgtm :+1:

File: packages/common/test/pipes/async_pipe_spec.ts (hunk 1 above)
Comment (crisbeto) [reply packages/common/test/pipes/async_pipe_spec.ts:25]: We'll have to make it when we eventually switch to rxjs 7. I figured I'd do it now since it's in a test.

File: integration/typings_test_rxjs7/tsconfig.json (hunk 3 above)
Comment (crisbeto) [reply integration/typings_test_rxjs7/tsconfig.json:24]: Not sure, I was following the same pattern as the TS tests.

File: packages/core/test/acceptance/property_binding_spec.ts [hunk 4]
```diff
@@ -187,7 +187,7 @@ describe('property bindings', () => {
     })
     class OtherDir {
       @Input() id: number|undefined;
-      @Output('click') clickStream = new EventEmitter();
+      @Output('click') clickStream = new EventEmitter<void>();
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/property_binding_spec.ts:190]: Just curious how break-y that would be for apps code as well and whether we should consider doing a migration (schematic) to simplify the upgrade? That also reminds me of #38916...

**Review by AndrewKushnir — APPROVED** [review]
LGTM, thanks @crisbeto 👍 

I've just left a comment about `EventEmitter`. This is not a problem for now since RxJS v7 is an opt-in, but we'd probably need to look at it when we decide to drop RxJS v6 support (in favor of v7).

File: packages/core/test/acceptance/property_binding_spec.ts (hunk 4 above)
Comment (crisbeto) [reply packages/core/test/acceptance/property_binding_spec.ts:190]: It'll be pretty break-ey, even for our own code, but as you mentioned, it's not a problem for now because version 7 is opt in. I explored it a bit in https://github.com/angular/angular/pull/41793 which revealed a few other breaking changes.

File: packages/common/test/pipes/async_pipe_spec.ts (hunk 1 above)
Comment (gkalpak) [reply packages/common/test/pipes/async_pipe_spec.ts:25]: Yeah, but why is this change necessary for RxJS v7? Aren't `input.subscribe(...args)` and `input.subscribe.apply(input, args)` equivalent?

Comment (crisbeto) [reply packages/common/test/pipes/async_pipe_spec.ts:25]: I honestly don't remember since it's been a while from when I tried updating it. I just know that it was a compilation error.

Comment (gkalpak) [reply packages/common/test/pipes/async_pipe_spec.ts:25]: I was curious, so I tried it out. The error is:
`A spread argument must either have a tuple type or be passed to a rest parameter.`

Alternatively, this could be fixed by changing the type of `args` (but `.apply()` is fine as well, I guess):

```diff
-        subscribe(...args: any): Unsubscribable {
+        subscribe(...args: Parameters<Subscribable<T>['subscribe']>): Unsubscribable {
           const subscription = input.subscribe(...args);
```

**Review by IgorMinar — APPROVED** [review]
This looks good to me. With all the aio tests running on rxjs 7 would should have enough test coverage to know when things break.

It's going to be really interesting to see how easy it will be for everyone to get from 6 to 7 and collect pain points that we might need to help or resolve before removing v6 support.

Thanks for getting this thought the line.

Reviewed-for: global-approvers
